File: MapsSDKExamplesObjc/Search/SearchChargingStationsViewController.m
/**
 * Copyright (c) 2019 TomTom N.V. All rights reserved.
 *
 * This software is the proprietary copyright of TomTom N.V. and its
 * subsidiaries and may be used for internal evaluation purposes or commercial
 * use strictly subject to separate licensee agreement between you and TomTom.
 * If you are the licensee, you are only permitted to use this Software in
 * accordance with the terms of your license agreement. If you are not the
 * licensee then you are not authorised to use this software in any manner and
 * should immediately return it to TomTom N.V.
 */

#import "SearchChargingStationsViewController.h"
#import <TomTomOnlineSDKSearch/TomTomOnlineSDKSearch.h>
#import <TomTomOnlineSDKMaps/TomTomOnlineSDKMaps.h>
#import <TomTomOnlineSDKRouting/TomTomOnlineSDKRouting.h>
#import <MapKit/MapKit.h>

@interface SearchChargingStationsViewController () <TTAnnotationDelegate>
@property(nonatomic, strong) EVChargingStationService *service;
@property(nonatomic, strong) TTRoute *planner;

@end

@implementation SearchChargingStationsViewController

- (OptionsView *)getOptionsView {
    return [[OptionsViewSingleSelect alloc] initWithLabels:@[ @"Along the route", @"Fuzzy", @"Geometry" ] selectedID:-1];
}

- (void)setupInitialCameraPosition {
    [self.mapView centerOnCoordinate:[TTCoordinate AMSTERDAM_A10] withZoom:10];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.service = [[EVChargingStationService alloc] initWithApiKey:Key.Search];

    self.mapView.annotationManager.delegate = self;
    self.planner = [[TTRoute alloc] initWithKey:Key.Routing];
}

#pragma mark OptionsViewDelegate

- (void)displayExampleWithID:(NSInteger)ID on:(BOOL)on {
    [super displayExampleWithID:ID on:on];
    [self.progress show];

    [self.mapView.annotationManager removeAllAnnotations];
    [self.mapView.annotationManager removeAllOverlays];
    [self.mapView.routeManager removeAllRoutes];

    switch (ID) {
    case 2:
        [self searchGeometry];
        break;
    case 1:
        [self searchFuzzy];
        break;
    default:
        [self searchAlongTheRoute];
        break;
    }
}
- (void)draw:(TTFullRoute *)route {
    TTMapRoute *mapRoute = [TTMapRoute routeWithCoordinatesData:route withRouteStyle:TTMapRouteStyle.defaultActiveStyle imageStart:TTMapRoute.defaultImageDeparture imageEnd:TTMapRoute.defaultImageDestination];
    [self.mapView.routeManager addRoute:mapRoute];
}

- (void)searchForChargingStationsAlong:(TTFullRoute *)route {

    NSMutableArray<CLLocation *> *routeLocations = [@[] mutableCopy];
    for (NSValue *value in route.coordinatesData) {
        CLLocation *location = [[CLLocation alloc] init:[value MKCoordinateValue]];
        [routeLocations addObject:location];
    }
    __weak SearchChargingStationsViewController *weakSelf = self;
    [self.service searchWithRoute:routeLocations
                       completion:^(NSArray<ChargingStationDetails *> *_Nullable result, NSError *_Nullable error) {
                         SearchChargingStationsViewController *strongSelf = weakSelf;
                         if (result != NULL && strongSelf != NULL) {
                             [strongSelf searchCompletedWithResult:result];
                         }
                       }];
}

#pragma mark Examples

- (void)searchFuzzy {
    __weak SearchChargingStationsViewController *weakSelf = self;
    [self.service searchWithTopLeft:[TTCoordinate AMSTERDAM_TOP_LEFT]
                        bottomRight:[TTCoordinate AMSTERDAM_BOTTOM_RIGHT]
                         completion:^(NSArray<ChargingStationDetails *> *_Nullable result, NSError *_Nullable error) {
                           SearchChargingStationsViewController *strongSelf = weakSelf;
                           if (result != NULL && strongSelf != NULL) {
                               [strongSelf searchCompletedWithResult:result];
                           }
                         }];
}

- (void)searchAlongTheRoute {
    TTRouteQuery *query = [[TTRouteQueryBuilder createWithDest:[TTCoordinate HAARLEM] andOrig:[TTCoordinate AMSTERDAM]] build];
    __weak SearchChargingStationsViewController *weakSelf = self;

    [self.planner planRouteWithQuery:query
                   completionHandler:^(TTRouteResult *_Nullable result, TTResponseError *_Nullable error) {
                     SearchChargingStationsViewController *strongSelf = weakSelf;

                     if (result.routes.firstObject != NULL) {
                         TTFullRoute *route = result.routes.firstObject;
                         [strongSelf draw:route];
                         [strongSelf searchForChargingStationsAlong:route];
                     }
                   }];
}

- (void)searchGeometry {

    CLLocationCoordinate2D coordinates[13];
    coordinates[0] = CLLocationCoordinate2DMake(52.37874, 4.90482);
    coordinates[1] = CLLocationCoordinate2DMake(52.37664, 4.92559);
    coordinates[2] = CLLocationCoordinate2DMake(52.37497, 4.94877);
    coordinates[3] = CLLocationCoordinate2DMake(52.36805, 4.97246);
    coordinates[4] = CLLocationCoordinate2DMake(52.34918, 4.95993);
    coordinates[5] = CLLocationCoordinate2DMake(52.34016, 4.95169);
    coordinates[6] = CLLocationCoordinate2DMake(52.32894, 4.91392);
    coordinates[7] = CLLocationCoordinate2DMake(52.34048, 4.88611);
    coordinates[8] = CLLocationCoordinate2DMake(52.33953, 4.84388);
    coordinates[9] = CLLocationCoordinate2DMake(52.37067, 4.8432);
    coordinates[10] = CLLocationCoordinate2DMake(52.38492, 4.84663);
    coordinates[11] = CLLocationCoordinate2DMake(52.40011, 4.85058);
    coordinates[12] = CLLocationCoordinate2DMake(52.38995, 4.89075);

    TTPolygon *mapPolygon = [TTPolygon polygonWithCoordinates:coordinates count:13 opacity:1 color:[TTColor RedSemiTransparent] colorOutline:[TTColor RedSemiTransparent]];
    [self.mapView.annotationManager addOverlay:mapPolygon];

    __weak SearchChargingStationsViewController *weakSelf = self;

    NSMutableArray *locations = [[NSMutableArray alloc] init];

    for (NSUInteger i = 0; i < 13; i++) {
        CLLocation *location = [[CLLocation alloc] init:coordinates[i]];
        [locations addObject:location];
    }
    PolygonShape *shape = [[PolygonShape alloc] initWithLocations:locations];
    [self.service searchWithShape:@[ shape ]
                       completion:^(NSArray<ChargingStationDetails *> *_Nullable result, NSError *_Nullable error) {
                         SearchChargingStationsViewController *strongSelf = weakSelf;
                         if (result != NULL && strongSelf != NULL) {
                             [strongSelf searchCompletedWithResult:result];
                         }
                       }];
}

- (void)searchCompletedWithResult:(NSArray<ChargingStationDetails *> *)stations {
    [self.progress hide];
    for (ChargingStationDetails *station in stations) {
        ChargeStationDetailsAnnotation *annotation = [[ChargeStationDetailsAnnotation alloc] initWithInfo:station];
        [self.mapView.annotationManager addAnnotation:annotation];
    }
}

- (void)annotationManager:(id<TTAnnotationManager>)manager annotationSelected:(TTAnnotation *)annotation {
    [self.mapView centerOnCoordinate:annotation.coordinate];
}

- (UIView<TTCalloutView> *)annotationManager:(id<TTAnnotationManager>)manager viewForSelectedAnnotation:(TTAnnotation *)selectedAnnotation {
    ChargeStationDetailsAnnotation *annotation = (ChargeStationDetailsAnnotation *)selectedAnnotation;
    if (annotation == NULL) {
        return [[TTCalloutOutlineView alloc] initWithText:@"-"];
    } else {
        NSString *address = [annotation freeFormAddress];
        NSString *availabilityText = [annotation readableAvailability];
        NSString *text = [NSString stringWithFormat:@"%@\n%@", address, availabilityText];
        ChargingStationCalloutView *contentView = [[ChargingStationCalloutView alloc] initWithText:text title:annotation.info.name height:100];
        contentView.iconView.image = [UIImage imageNamed:@"ic_ev_slow_charging"];
        return [[TTCalloutOutlineView alloc] initWithUIView:contentView];
    }
}

@end

File: MapsSDKExamplesSwift/Search/SearchChargingStationsViewController.swift
//
/**
 * Copyright (c) 2020 TomTom N.V. All rights reserved.
 *
 * This software is the proprietary copyright of TomTom N.V. and its subsidiaries and may be used
 * for internal evaluation purposes or commercial use strictly subject to separate licensee
 * agreement between you and TomTom. If you are the licensee, you are only permitted to use
 * this Software in accordance with the terms of your license agreement. If you are not the
 * licensee then you are not authorised to use this software in any manner and should
 * immediately return it to TomTom N.V.
 */

import CoreLocation
import MapKit
import MapsSDKExamplesCommon
import MapsSDKExamplesVC
import TomTomOnlineSDKMaps
import TomTomOnlineSDKRouting
import TomTomOnlineSDKSearch
import UIKit

class SearchChargingStationsViewController: MapBaseViewController {
    let service = EVChargingStationService(apiKey: Key.Search)
    let routePlanner = TTRoute(key: Key.Routing)

    override func getOptionsView() -> OptionsView {
        return OptionsViewSingleSelect(labels: ["Along the route", "Fuzzy", "Geometry"], selectedID: -1)
    }

    override func setupInitialCameraPosition() {
        mapView.center(on: TTCoordinate.AMSTERDAM_A10(), withZoom: 10)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        mapView.annotationManager.delegate = self
    }

    private func drawShapes(coordinates: [CLLocationCoordinate2D]) {
        mapView.annotationManager.removeAllAnnotations()
        var coordinates = [CLLocationCoordinate2DMake(52.37874, 4.90482),
                           CLLocationCoordinate2DMake(52.37664, 4.92559),
                           CLLocationCoordinate2DMake(52.37497, 4.94877),
                           CLLocationCoordinate2DMake(52.36805, 4.97246),
                           CLLocationCoordinate2DMake(52.34918, 4.95993),
                           CLLocationCoordinate2DMake(52.34016, 4.95169),
                           CLLocationCoordinate2DMake(52.32894, 4.91392),
                           CLLocationCoordinate2DMake(52.34048, 4.88611),
                           CLLocationCoordinate2DMake(52.33953, 4.84388),
                           CLLocationCoordinate2DMake(52.37067, 4.8432),
                           CLLocationCoordinate2DMake(52.38492, 4.84663),
                           CLLocationCoordinate2DMake(52.40011, 4.85058),
                           CLLocationCoordinate2DMake(52.38995, 4.89075)]
        let mapPolygon = TTPolygon(coordinates: &coordinates, count: UInt(coordinates.count), opacity: 1, color: TTColor.RedSemiTransparent(), colorOutline: TTColor.RedSemiTransparent())
        mapView.annotationManager.add(mapPolygon)
    }

    // MARK: OptionsViewDelegate

    override func displayExample(withID ID: Int, on: Bool) {
        super.displayExample(withID: ID, on: on)
        progress.show()
        mapView.annotationManager.removeAllAnnotations()
        mapView.annotationManager.removeAllOverlays()
        mapView.routeManager.removeAllRoutes()

        switch ID {
        case 2:
            searchGeometry()
        case 1:
            searchFuzzy()
        default:
            searchAlongTheRoute()
        }
    }

    // MARK: Examples

    func searchFuzzy() {
        service.search(topLeft: TTCoordinate.AMSTERDAM_TOP_LEFT(), bottomRight: TTCoordinate.AMSTERDAM_BOTTOM_RIGHT()) { result, _ in
            guard let result = result else {
                return
            }
            self.search(completedWith: result)
        }
    }

    func searchAlongTheRoute() {
        let query = TTRouteQueryBuilder.create(withDest: TTCoordinate.HAARLEM(), andOrig: TTCoordinate.AMSTERDAM())
            .build()

        routePlanner.plan(with: query) { [weak self] result, _ in
            guard let route = result?.routes.first else {
                return
            }

            let mapRoute = TTMapRoute(coordinatesData: route,
                                      with: TTMapRouteStyle.defaultActive(),
                                      imageStart: TTMapRoute.defaultImageDeparture(),
                                      imageEnd: TTMapRoute.defaultImageDestination())
            self?.mapView.routeManager.add(mapRoute)

            let routeLocations = route.coordinatesData().map { CLLocation($0.mkCoordinateValue) }
            self?.service.search(route: routeLocations, completion: { chargingStationResult, _ in
                guard let chargingStationResult = chargingStationResult else {
                    return
                }
                self?.search(completedWith: chargingStationResult)
            })
        }
    }

    func searchGeometry() {
        let coordinates = [CLLocationCoordinate2DMake(52.37874, 4.90482),
                           CLLocationCoordinate2DMake(52.37664, 4.92559),
                           CLLocationCoordinate2DMake(52.37497, 4.94877),
                           CLLocationCoordinate2DMake(52.36805, 4.97246),
                           CLLocationCoordinate2DMake(52.34918, 4.95993),
                           CLLocationCoordinate2DMake(52.34016, 4.95169),
                           CLLocationCoordinate2DMake(52.32894, 4.91392),
                           CLLocationCoordinate2DMake(52.34048, 4.88611),
                           CLLocationCoordinate2DMake(52.33953, 4.84388),
                           CLLocationCoordinate2DMake(52.37067, 4.8432),
                           CLLocationCoordinate2DMake(52.38492, 4.84663),
                           CLLocationCoordinate2DMake(52.40011, 4.85058),
                           CLLocationCoordinate2DMake(52.38995, 4.89075)]

        drawShapes(coordinates: coordinates)

        let locations = coordinates.map { CLLocation(latitude: $0.latitude, longitude: $0.longitude) }
        let polygonShape = PolygonShape(locations: locations)
        service.search(shape: [polygonShape]) { result, _ in
            guard let result = result else {
                return
            }
            self.search(completedWith: result)
        }
    }

    func search(completedWith stations: [ChargingStationDetails]) {
        progress.hide()
        for result in stations {
            let annotation = ChargeStationDetailsAnnotation(info: result)
            mapView.annotationManager.add(annotation)
        }
    }
}

extension SearchChargingStationsViewController: TTAnnotationDelegate {
    func annotationManager(_: TTAnnotationManager, annotationSelected annotation: TTAnnotation) {
        mapView.center(on: annotation.coordinate)
    }

    func annotationManager(_: TTAnnotationManager, viewForSelectedAnnotation selectedAnnotation: TTAnnotation) -> UIView & TTCalloutView {
        guard let selectedAnnotation = selectedAnnotation as? ChargeStationDetailsAnnotation else {
            return TTCalloutOutlineView(text: "-")
        }
        let address = selectedAnnotation.freeFormAddress()
        let availabilityText = selectedAnnotation.readableAvailability()
        let text = "\(address)\n\(availabilityText)"
        let view = ChargingStationCalloutView(text: text, title: selectedAnnotation.info.name, height: 100)
        view.iconView.image = UIImage(named: "ic_ev_slow_charging")
        return TTCalloutOutlineView(uiView: view)
    }
}
